### ISSUE TITLE
Improvement to the offset highlighting in text-view 

### ISSUE THREAD

**Issue by sebastian-nehrdich** [ayya-vimala/pali-networks#83]
@ayya-vimala as I said in the other issue, I have the impression that the offset highlighting is now working well and we have done a good job (you particularly!) 
In my communication with M we could already figure out two simple changes on the frontend (which however are probably big changes on the inside) that could improve the usability of the website dramatically. These are:
1. when a user clicks on a segment on the left side, we should actually capture the mouse-cursor-position and display only the matches in the middle column where a part of the parallel lies within the region of the click on the root-segment. I know that implementing this will be another visit to the lower realms and I am not particularly excited to go there but it will be for the benefit of the users!
2. As one hovers the mouse over the parallels in the middle column, the background highlighting in the left side (which is currently yellow after one has clicked a certain segment) should change to reflect  the size of the root-segments (respecting the root_offset_values) that are associated with the current parallel under the mouse cursor. I have implemented something similar before when I worked on the Sanskrit-Tibetan-parallel view based on K's code. I expect the implementation of this to be even worse than (1) but again it would be for the benefit of all the beings (利樂諸有情)! 

Just to make sure, I am not expecting you to implement this, but I think it's important to discuss this together and see how we can tackle this! I think I should be able to implement this, but when and how is another question. 
We can discuss this also more in detail tomorrow in the meeting. I just thought to note this down before I forget about the details again.

**Comment by ayya-vimala** [comment]
I like your expression that implementing this might be another visit to the lower realms :-)
I'm also not sure this would be such a good thing because I think it has the potential to cause a lot of confusion for users too. It would mean that if you click on a segment, it gives different results each time. But anyway, not only would it require to register where the mouse is clicked (which in itself is rather simple) but also which word happens to be at that spot on the screen and then figure out where that word is in relation to the segment. In any case, this really is beyond my capabilities. Maybe if they put in a lot of money to hire a good developer ...

With number 2: seems to me that this is already done ??? or have I got something wrong.

I think we really have to focus on what is most important and not get lost in smaller details like this.

**Comment by sebastian-nehrdich** [comment]
Unfortunately I received a lot of criticism from O because of the way we
display the matches currently. But she also didn't give any ideas on how to
improve things yet... Let's see.

**Comment by ayya-vimala** [comment]
As I understood from our discussion today, it was agreed that:
1. Yes, it would be good if the leftside text displays all the parallel quotes with the right-side text and ideally becomes clickable.
2. Will wait with this issue until queries are possible with a database.

**Comment by sebastian-nehrdich** [comment]
I think (1) should be the other way around, so right text should display
all the parallel quotes with the left text (and ideally clickable). The
left text is not a problem, it's working good already.
(2) is right!

**Comment by ayya-vimala** [comment]
Yes, absolutely right. Thanks for pointing that out.

**Comment by sebastian-nehrdich** [comment]
Just to warn you and prepare you for the shock, I started last night to rewrite the entire offset highlighting code because it's very necessary for my upcoming presentation. I am about halfway through now and might upload it to a new branch today...

**Comment by ayya-vimala** [comment]
If you let me know what exactly you want then I can do it and then you can focus on getting those files onto the server. I know that's probably a pretty boring job but I cannot get on unless I have some example files.

**Comment by sebastian-nehrdich** [comment]
(new Tibetan files are online)
Well what we need to change is a bit more complex! I have now implemented
the highlighting in a way that it is sensitive to the word that one clicks
on (so only those matches show up where the current word under the cursor
lies within root_offset_beg and root_offset_end). Needs a bit of debugging
here and there but basicly works. Currently I am adding the function that
as one moves the mouse over the middle column, the whole size of the match
with precise boundaries based on offset-values for that current match gets
highlighted on the left side. I am also making progress on that and think
that I can wrap this up today.
These are two complex changes but I got grilled heavily for that by
reviewers the last week and I am afraid that if we don't get it right
barely anybody would like to use our website later...

**Comment by ayya-vimala** [comment]
You might find when doing this you get a lot of bugs, also in the segment-view.

**Comment by ayya-vimala** [comment]
Is that why you need these things?

```
                "quote_position_beg": 21098425,
                "quote_position_last": 21098425,
                "head_beg": 21045583,
                "head_end": 21045589,
```

**Comment by sebastian-nehrdich** [comment]
Nono it has nothing to do with these values (they are just there fore
debugging purposes).

**Comment by sebastian-nehrdich** [comment]
Ok it is now more or basicly working, give and take some bugs that I didn't hunt down yet. 
The branch is new-offset-highlighting. 
Each word get's wrapped into a span-box and has a position-value so we can modify the highlighting of each word by querying the document-tree. This is a solution to the problem of precise highlighting and precise linking, but it breaks many things of course. 
For example at the moment the segment-view and all the cleanSegment-code is broken and rewritting will need a bit of patience to make it work with the new layout. 
Generally speaking, I was not excited to make these changes but they are not a question of nice to have but essential to the usability of this tool as a research-tool, so I was told quite clearly that this is not going to be accepted if these things are not going to work out well. Of course regarding the styling and the concrete solution of the highlighting, I am totally open to your suggestions. It looks quite terrible now, but I am after more than 10 hours of programming today already quite done with this and can't do much more.

**Comment by sebastian-nehrdich** [comment]
forgot to mention, Chinese is now broken, Sanskrit certainly as well...

**Comment by ayya-vimala** [comment]
Well yes, I'm sure this is a very difficult thing to do and it will break many things. That is why I was not keen to do it until we have a database. 
I'll try to merge master but there are many conflicts.

**Comment by ayya-vimala** [comment]
Okay, I have integrated master into your branch so you have to pull before doing anything.
I have also cleaned up the code (did not remove the console.logs which you need to do at the end).
And I could not stand the layout so I have cleaned up the words a bit so it looks a bit more as it should be.

Main problem now I see that as soon as you move over the middle column text, the highlight in the right disappears. (apart from the various languages and segment view not working). I have not looked into this because I don't feel I want to mess around with your code because that will only create more difficulties for you.

**Comment by sebastian-nehrdich** [comment]
Thank you for this! 
Yes I am sure there are problems enough and we should collect them and see what we can do with them. DB will certainly help with simplifying the datatransfer (no need to 'misuse' poor span attributes for that)  but I think separating tokens in the DOM would have been necessary also without a DB. Regarding the way things are done on the display-level I am totally open to your suggestions. For the time being it was most important for me to get control over each token of the root text seperately in the DOM so I didn't think much yet about the handling/display. 
I was thinking as a next step to do the highlighting of the passages based on the co-occ-value with a color gradient. So passages where we only have a quote with one co-occ get highlighted in red and passages with many co-occs (maybe '10 or more') get highlighted in blue. and everything between that will be on the gradient between these colors depending on how high the co-occ value is. This should also help a lot with the Pāli so one get's a color feedback whether a passage is a bulk-quote or not and can focus on important parallels more quickly.

**Comment by sebastian-nehrdich** [comment]
(of course red and blue are just quick guesses, I am very bad at design/color and maybe we can have a much nicer layout!) 
And also sorry for the source code formatting. I am quite a fan of js2-mode for emacs but it seems that it just doesn't like very much the whitespace-format of your files and it's a constant battle to not mess things up entirely.

**Comment by ayya-vimala** [comment]
js automatic formatting on text editors does not work well with webcomponents. There is another syntax for that but I never bothered installing it myself.

**Comment by sebastian-nehrdich** [comment]
I have just pushed a slightly updated version to the new branch (based on your merge). It now also has a rudimentary function of calculating the color gradients before marking the text on the left side. The idea is to give frequent quotes a light color (= not that important) and more rare quotes a darker color (= important). I am sure we can improve a lot on the display, but technically it kind of works and is not doing too bad I think.

**Comment by sebastian-nehrdich** [comment]
I did another update, this time I made major changes to the segment-view code so we can implement sorting of the results. I wrote a bit more on that here: https://github.com/ayya-vimala/pali-networks/issues/45#issuecomment-518063998_
Also I tuned the performance a bit by reducing the calls to html`` (seems that these become quite expensive if we do more than a few thousand per file...).
Rendering time of the text-view is still a bit high though for longer files. Certainly we can do a bit better here, but the basic functionality now is all there!
